### PR TITLE
module added for schedule 

### PR DESCRIPTION
**Pull request by nirupaman**
root@cssosbe03-b12:/home/nirun/niru_ansiblecopy/hpe3par_ansible_module# ansible-playbook schedule.yml
 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

 [WARNING]: While constructing a mapping from /home/nirun/niru_ansiblecopy/hpe3par_ansible_module/remote_copy.yml, line 3,
column 5, found a duplicate dict key (http_proxy). Using last defined value only.

PLAY [localhost] ************************************************************************************************************

TASK [Gathering Facts] ******************************************************************************************************
ok: [localhost]

TASK [Create Schedule on source] ********************************************************************************************
changed: [localhost]

TASK [pause] ****************************************************************************************************************
[pause]
Press Enter:
ok: [localhost]

TASK [Delete Schedule from source] ******************************************************************************************
changed: [localhost]

PLAY RECAP ******************************************************************************************************************
localhost                  : ok=4    changed=2    unreachable=0    failed=0

### CODE REVIEW

**Comment by nirupaman** [comment]
@farhan7500 Please review

**Comment by farhan7500** [comment]
Can you also paste the CLI output showing the schedule has been created

File: Modules/hpe3par_snapshot.py [hunk 1]
```diff
@@ -435,13 +447,131 @@ def restore_snapshot_online(
...
+            False,
+            "Schedule create failed. Base volume name is null",
+            {})
+    if len(base_volume_name) < 1 or len(base_volume_name) > 31:
+        return (False, False, "Schedule create failed. Base volume name must be atleast 1 character and not more than 31 characters", {})
+    try:
+        client_obj.login(storage_system_username, storage_system_password)
+        client_obj.setSSHOptions(storage_system_ip, storage_system_username, storage_system_password)
+        expirationHours = convert_to_hours(expiration_time, expiration_unit)
+	retentionHours = convert_to_hours(retention_time, retention_unit)
```
Comment (farhan7500) [inline Modules/hpe3par_snapshot.py]: Is this indented correctly?

File: Modules/hpe3par_snapshot.py [hunk 2]
```diff
@@ -435,13 +447,131 @@ def restore_snapshot_online(
...
+        if client_obj.volumeExists(base_volume_name):
+           cmd = ["\"createsv"]
```
Comment (farhan7500) [inline Modules/hpe3par_snapshot.py]: What happens when the schedule that we are creating already exists? We need to check if a schedule is created by that name. Therefore, the following methods need to be added in SDK and client

a. getSchedule(name)
b. scheduleExists(name)

scheduleExists will refer try and get the schedule and based on that result we have to decide in ansible whether to create the new schedule

File: Modules/hpe3par_snapshot.py [hunk 3]
```diff
@@ -435,13 +447,131 @@ def restore_snapshot_online(
...
+            False,
+            "schedule delete failed. Schedule name is null",
+            {})
+    if len(schedule_name) < 1 or len(schedule_name) > 31:
+        return (False, False, "schedule create failed. Schedule name must be atleast 1 character and not more than 31 characters", {})
+    try:
+        client_obj.login(storage_system_username, storage_system_password)
+        client_obj.setSSHOptions(storage_system_ip, storage_system_username, storage_system_password)
+        if client_obj.scheduleExists(schedule_name):
+           client_obj.deleteSchedule(schedule_name)
```
Comment (farhan7500) [inline Modules/hpe3par_snapshot.py]: Same again. What happens if the schedule does not exist and we try to delete it

File: Modules/hpe3par_snapshot.py (hunk 2 above)
Comment (nirupaman) [reply Modules/hpe3par_snapshot.py]: If schedule exist it show error that schedule already exist ..Can we have only one function scheduleExists(name) in sdk

File: Modules/hpe3par_snapshot.py (hunk 3 above)
Comment (nirupaman) [reply Modules/hpe3par_snapshot.py]: handled

File: Modules/hpe3par_snapshot.py [hunk 4]
```diff
@@ -117,11 +117,21 @@
     description:
       - "Specifies a snapshot volume name."
     required: true
+  schedule_name:
+    description:
+      - "Name of the schedule."
+    required: true
+  task_freq:
+    description:
+      - "Frequency of the schedule to be created."
```
[the remark this answers is not in the document]
Comment (nirupaman) [reply Modules/hpe3par_snapshot.py]: added option for task_freq

**Comment by nirupaman** [comment]
Success Output

1) pytest run
root@cssosbe03-b12:/home/nirun/niru_ansiblecopy/hpe3par_ansible_module# pytest test/unit/test_hpe3par_snapshot.py
=====================  test_hpe3par_snapshot.py  =====================
...............
----------------------------------------------------------------------
Ran 15 tests in 0.025s

OK
*******************************************************************************

2) Playbook create and delete schedule

Create Schedule

root@cssosbe03-b12:/home/nirun/niru_ansiblecopy/hpe3par_ansible_module# ansible-playbook playbooks/create_sch_playbook.yml
 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

 [WARNING]: While constructing a mapping from
/home/nirun/niru_ansiblecopy/hpe3par_ansible_module/playbooks/create_sch_playbook.yml, line 3, column 5, found a duplicate
dict key (http_proxy). Using last defined value only.


PLAY [localhost] ************************************************************************************************************

TASK [Gathering Facts] ******************************************************************************************************
ok: [localhost]

TASK [Create Schedule on source] ********************************************************************************************
changed: [localhost]

PLAY RECAP ******************************************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0

root@cssosbe03-b12:/home/nirun/niru_ansiblecopy/hpe3par_ansible_module#
==============================================================
Delete Schedule

root@cssosbe03-b12:/home/nirun/niru_ansiblecopy/hpe3par_ansible_module# ansible-playbook playbooks/delete_sch_playbook.yml
 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

 [WARNING]: While constructing a mapping from
/home/nirun/niru_ansiblecopy/hpe3par_ansible_module/playbooks/delete_sch_playbook.yml, line 3, column 5, found a duplicate
dict key (http_proxy). Using last defined value only.


PLAY [localhost] ************************************************************************************************************

TASK [Gathering Facts] ******************************************************************************************************
ok: [localhost]

TASK [Delete Schedule from source] ******************************************************************************************
changed: [localhost]

PLAY RECAP ******************************************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0

File: Modules/hpe3par_snapshot.py [hunk 5]
```diff
@@ -505,7 +637,18 @@ def main():
         },
         "rm_exp_time": {
             "type": "bool"
+        },
+        "schedule_name": {
+            "type": "str"
+        },
+        "task_freq": {
+            "type": "str",
+            "choices": ['yearl', 'monthly', 'weekly', 'daily', 'hourly']
```
Comment (farhan7500) [inline Modules/hpe3par_snapshot.py]: This does not take into account cron type option

And I guess it must be yearl'y'

File: Modules/hpe3par_snapshot.py [hunk 6]
```diff
@@ -435,13 +447,134 @@ def restore_snapshot_online(
...
+            "Schedule create failed. Base volume name is null",
+            {})
+    if len(base_volume_name) < 1 or len(base_volume_name) > 31:
+        return (False, False, "Schedule create failed. Base volume name must be atleast 1 character and not more than 31 characters", {})
+    try:
+        client_obj.login(storage_system_username, storage_system_password)
+        client_obj.setSSHOptions(storage_system_ip, storage_system_username, storage_system_password)
+        expirationHours = convert_to_hours(expiration_time, expiration_unit)
+	retentionHours = convert_to_hours(retention_time, retention_unit)
+        if client_obj.volumeExists(base_volume_name):
```
[the remark this answers is not in the document]
Comment (nirupaman) [reply Modules/hpe3par_snapshot.py]: We also need to check if volume exists

File: Modules/hpe3par_snapshot.py [hunk 7]
```diff
@@ -435,13 +447,134 @@ def restore_snapshot_online(
...
+           cmd = ["createsv"] 
+	   if read_only:
+              cmd.append("-ro")
```
[the remark this answers is not in the document]
Comment (nirupaman) [reply Modules/hpe3par_snapshot.py]: which value?

**Comment by nirupaman** [comment]
@farhan7500 : Please review added scheduleExists and try except in getSchedule

File: Modules/hpe3par_snapshot.py (hunk 5 above)
Comment (nirupaman) [reply Modules/hpe3par_snapshot.py]: changed

**Comment by farhan7500** [comment]
Please remove ALL unrelated changes from the PR. Only then will I be able to review properly

**Comment by nirupaman** [comment]
Duplicate

**Pull request closed without merging**